src/backend/metal/src/device.rs

```diff
@@ -114,10 +114,9 @@ fn get_final_function(
     })?;
 
     if !function_specialization {
-        assert!(
-            specialization.data.is_empty() && specialization.constants.is_empty(),
-            "platform does not support specialization",
-        );
+        if specialization.data.is_empty() && specialization.constants.is_empty() {
```

Comment:
the condition should be the opposite: if specialization has anything, we should be erroring out

Reply:
Could also return a `FunctionError` btw, which is slightly better, but I'm fine either way, since it's not the focus of your PR